### ISSUE TITLE
Invalid type inference for optional chaining operator when nullish check is extracted to `const` varaiable

### ISSUE THREAD

**Issue by jet2jet** [microsoft/TypeScript#58632]
### 🔎 Search Terms

"optional chaining operator", null check

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (from 4.4.4 to 5.5.0-dev.20240523)

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.5.0-dev.20240523#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChkHJwBcyAzmFKAObIA+yIArgDYsDcuAvrjEyAjDB0IZACM4UABQxMpDFgbM2AShz5CCERWTAyAOVYs9AC2QBeZLPQB+AHRwLl5Rw0FgMZFL2G2ptXiEQchQEGBMUCCcQTxBAPRxIRAADuhQYMgARADyYgBWEIK6ZMipZGTAYiwAnsgA5C51dlIATACsAMwAjCqZVmlZ1g6ZbshaIGToLBB2LOjUMpgOdmDoAKrJydAAwnBkEFIqKpw8QA

### 💻 Code

```ts
interface Foo {
    a: string | null;
}
function bar(foo: Foo | null) {
    const isNullish = foo?.a == null;
    if (isNullish) {
        return;
    }
    // report "Object is possibly 'null'.(2531)" for "foo.a"
    console.log(foo.a.toUpperCase());
}
```

### 🙁 Actual behavior

Report an error `Object is possibly 'null'.(2531)` for `foo.a`

### 🙂 Expected behavior

No error is reported

### Additional information about the issue

I don't know this behavior is a bug or intended, but based on PR #44730 , I chose 'bug' for this issue.

**Comment by whzx5byb** [comment]
Per https://github.com/microsoft/TypeScript/pull/44730:
> Narrowing through indirect references occurs only when the conditional expression or discriminant property access is declared in a const variable declaration with no type annotation, and the reference being narrowed is a const variable, **a readonly property**, or a parameter for which there are no assignments in the function body.

So just mark the property `a` as readonly and it works.
```ts
interface Foo {
    readonly a: string | null;
}
```

**Comment by jet2jet** [comment]
Thanks for the response. I missed that this is applied only for 'a readonly property'.  
I'll close the issue.
